proposals/0000-language-editions.rst

```diff
@@ -0,0 +1,1144 @@
+  files. If a file is being loaded and does not specify its own language
+  edition, it inherits from the language edition in force at the top-level.
+
+- A language edition can control almost all behaviors of GHC. The meaning
+  (or existence) of other flags can depend on language edition. While
+  we will not implement it this way, we can imagine that GHC becomes
+  a set of programs that happen to share a binary; the choice of which
+  program is chosen by the language edition.
+
+  The one restriction on the expressive power of language editions
+  is that build products of different language editions (within the
+  same version of GHC) must be
+  compatible. We expect the Haskell ecosystem to contain packages
+  compiled with a variety of language editions, and they must work
+  together. The word *compatible* above is doing some heavy lifting,
+  in that it contains a notion that the interface (i.e. exported
+  symbols, their types, etc.) can be translated from one language
+  edition to another. This aspect of Haskell's design has always
+  been present, in that different files can be compiled with different
+  langauge extensions. We often don't notice this, because we have
+  been careful with language extensions not to make the translation
+  apparent.
+
+  Although language editions have wide authority, we must be tasteful
+  in how they work. It would be problematic to have ``Stable2024``
+  and ``Stable2027`` disagree on the meaning of widely used features.
+  Yet I think it's best to be maximally expressive here, relying on
+  our future selves not to abuse our power.
+
+  (In the conversation, it was suggested that a ``Python`` language
+  edition would be beyond reason... but actually I think such a thing
+  would be lovely, if the necessary work was done to translate the
+  interfaces between Python and Haskell.)
+
+- For backward compatibility, a language edition can be specified
+  at an arbitrary place in a command-line invocation of GHC, or in a
+  ``LANGUAGE`` pragma in a file. Even though it might come later in
+  a command line, the edition can affect the meaning of command-line
+  arguments that precede it.
+
+- A cabal file will allow a new
+  field ``language-edition``, available both at top-level and in
+  build-product stanzas. This will specify the language edition. To
+  support backward compatibility, this will use the ``default-language``
+  setting if that is available, and omitting the ``language-edition`` will
+  use the default. At some point, it is expected that ``language-edition``
+  will become required.
+
+- Any new language features invented once e.g. ``Stable2027`` is released
+  will not be available with the 2024 editions. That is, if we introduce a new
+  feature ``-XDependentTypes`` in 2028, then enabling ``-XDependentTypes``
+  with ``Stable2024`` (or even ``Experimental2024``) will be an error.
+  This policy encourages users to upgrade their editions in order
+  to access GHC's new features.
```

Comment:
Hmmm.... interesting point. Yes I think it would be nice to allow this. Specifically, I could see allowing a language-edition override if the overriding language edition is a later version of the same series. In your proposed scenario, the cabal file would say `Stable2024` (meaning: this package will continue to compile until at least 2030), but individual modules could have `Stable2027` (meaning: this package will continue to compile until at least 2033). This works because the stability guarantee of `Stable2027` is strictly stronger than that of `Stable2024`. It doesn't work as nicely for overriding a `Stable2024` promise in a cabal file with an `Experimental` module.

How does this sound? I went to just incorporate this into the proposal, but it's not all that easy to do so, as doing it right probably requires having a notion of `Stable` separate from `2024`. (This separation has been suggested elsewhere, too, but I didn't see what new flexibility / expressiveness the distinction granted -- until now.) I'll leave this as an underspecified Alternative for now, but if we end up settling in agreement here, I'll rewrite the proposal accordingly.